### PR TITLE
Azure LUIS Support

### PR DESCRIPTION
**Pull request by thegoldenmule**
This is an experimental prototype with accompanying preview JS API.

- ISpeechController interface, distinct from voice input.
- Azure cognitive services SDK.

### CODE REVIEW

File: Assets/Source/Input/Speech/AzureSpeechController.cs [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+        /// <summary>
+        /// Called when an intent has been canceled.
+        /// </summary>
+        private void Intent_OnCanceled(object sender, IntentRecognitionCanceledEventArgs e)
+        {
+            //Log.Info(this, "OnCanceled.");
+
+            if (e.Reason == CancellationReason.Error)
+            {
+                Log.Info(this, "There was a error: {0}", e.ErrorDetails);
```
Comment (jsundra) [inline Assets/Source/Input/Speech/AzureSpeechController.cs]: Should this log as an Error, or does it proc over-zealously?

File: Assets/Source/Input/Speech/AzureSpeechController.cs [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+            //Log.Info(this, $"OnSpeechEndDetected: {e.Offset}");
+        }
+
+        /// <summary>
+        /// Called when speech has started.
+        /// </summary>
+        private void Intent_OnSpeechStartDetected(object sender, RecognitionEventArgs e)
+        {
+            //Log.Info(this, "OnSpeechStartDetected.");
+        }
```
Comment (jsundra) [inline Assets/Source/Input/Speech/AzureSpeechController.cs]: I assume these Intent callbacks were during original development for debugging? Can they be removed?

File: Assets/Source/Player/Scripting/Interface/App/RandomJsApi.cs [hunk 3]
```diff
@@ -0,0 +1,55 @@
+﻿using UnityEngine;
+
+namespace CreateAR.EnkluPlayer.Scripting
+{
+    [JsInterface("random-preview")]
+    public class RandomJsApi
```
Comment (jsundra) [inline Assets/Source/Player/Scripting/Interface/App/RandomJsApi.cs]: Missing class comment

File: Assets/Source/Input/Speech/AzureSpeechController.cs (hunk 1 above)
Comment (thegoldenmule) [reply Assets/Source/Input/Speech/AzureSpeechController.cs]: Consider `Error` as "Should I get an email about it?" In this case, no. It's similar to touch input events on an iPhone. There are valid reasons a cancel can be called, none of them catastrophic.

File: Assets/Source/Player/Scripting/Interface/App/RandomJsApi.cs (hunk 3 above)
Comment (thegoldenmule) [reply Assets/Source/Player/Scripting/Interface/App/RandomJsApi.cs]: Forgot about this guy.

File: Assets/Source/Input/Speech/AzureSpeechController.cs [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+            Log.Info(this, "OnSessionStopped.");
+        }
+
+        /// <summary>
+        /// Called when an intent is being recognized.
+        /// </summary>
+        private void Intent_OnRecognizing(object sender, IntentRecognitionEventArgs e)
+        {
+            //Log.Info(this, "OnRecognizing.");
+        }
```
[the remark this answers is not in the document]
Comment (thegoldenmule) [reply Assets/Source/Input/Speech/AzureSpeechController.cs]: Agreed!

**Pull request merged**